SimPEG/directives/pgi_directives.py

```diff
@@ -27,108 +26,142 @@
-                    tol=self.pgi_reg.gmm.tol,
-                    verbose_interval=self.pgi_reg.gmm.verbose_interval,
-                    warm_start=self.pgi_reg.gmm.warm_start,
-                    fixed_membership=self.fixed_membership,
-                )
-                clfupdate = clfupdate.fit(model)
+        """Run after the end of each iteration in the inversion."""
+        # Don't update the model if we are in the first iteration or if the
+        # current iteration doesn't match the update rate
+        if self.opt.iter == 0 or self.opt.iter % self.update_rate != 0:
+            return None
 
+        # Get current model
+        m = self.invProb.model
+        modellist = self.pgi_regularization.wiresmap * m
+        model = np.c_[
+            [a * b for a, b in zip(self.pgi_regularization.maplist, modellist)]
+        ].T
+
+        # Update the GMM parameters if required
+        gmmref = self.pgi_regularization.gmmref
+        if self.update_gmm:
+            # Determine the class of the new gmm
+            if isinstance(gmmref, GaussianMixtureWithNonlinearRelationships):
+                new_gmm_class = GaussianMixtureWithNonlinearRelationshipsWithPrior
+            elif isinstance(gmmref, WeightedGaussianMixture):
+                new_gmm_class = GaussianMixtureWithPrior
             else:
-                clfupdate = copy.deepcopy(self.pgi_reg.gmmref)
-
-            self.pgi_reg.gmm = clfupdate
-            membership = self.pgi_reg.gmm.predict(model)
-
-            if self.fixed_membership is not None:
-                membership[self.fixed_membership[:, 0]] = self.fixed_membership[:, 1]
+                raise ValueError(
+                    f"Invalid class '{gmmref.__class__}' for the Gaussian Mixture"
+                    " in the PGI regularization."
+                )
+            new_gmm = new_gmm_class(
+                gmmref=gmmref,
+                zeta=self.zeta,
+                kappa=self.kappa,
+                nu=self.nu,
+                verbose=self.verbose,
+                prior_type="semi",
+                update_covariances=self.update_covariances,
+                max_iter=self.pgi_regularization.gmm.max_iter,
+                n_init=self.pgi_regularization.gmm.n_init,
+                reg_covar=self.pgi_regularization.gmm.reg_covar,
+                weights_init=self.pgi_regularization.gmm.weights_,
+                means_init=self.pgi_regularization.gmm.means_,
+                precisions_init=self.pgi_regularization.gmm.precisions_,
+                random_state=self.pgi_regularization.gmm.random_state,
+                tol=self.pgi_regularization.gmm.tol,
+                verbose_interval=self.pgi_regularization.gmm.verbose_interval,
+                warm_start=self.pgi_regularization.gmm.warm_start,
+                fixed_membership=self.fixed_membership,
+            )
+            new_gmm = new_gmm.fit(model)
+        else:
+            new_gmm = copy.deepcopy(gmmref)
```

Comment:
I don't see the point of  creating a deep copy of the GMM after each iteration in case that the GMM is not intended to be updated. It's a waste of memory and time. Is there any reason why we should be doing it that I'm not seeing?

Reply:
I think I had funky behaviors back in the days when not creating a new object. But sklearn has deeply changed since so it might not be necessary anymore. Try it :)